### PR TITLE
[GSK 2658] Custom scanner

### PR DESCRIPTION
**Pull request by bmalezieux**
## Description

Add an example manager to deal with new data types in the scan report. The example manager should be able to add examples and render html code to be integrated into the report.

## Type of Change

- [ ] 📚 Examples / docs / tutorials / dependencies update
- [ ] 🔧 Bug fix (non-breaking change which fixes an issue)
- [ ] 🥂 Improvement (non-breaking change which improves an existing feature)
- [x] 🚀 New feature (non-breaking change which adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🔐 Security fix

## Checklist

- [x] I've read the [`CODE_OF_CONDUCT.md`](https://github.com/Giskard-AI/ai-inspector/blob/master/CODE_OF_CONDUCT.md) document.
- [x] I've read the [`CONTRIBUTING.md`](https://github.com/Giskard-AI/ai-inspector/blob/master/CONTRIBUTING.md) guide.
- [x] I've updated the code style using `make codestyle`.
- [x] I've written tests for all new methods and classes that I created.
- [x] I've written the docstring in Google format for all the methods and classes that I used.

### CODE REVIEW

**Comment by rabah-khalek** [comment]
Could you please also add on this branch a way to disable the "warning" and "Debug your issues in the Giskard hub" parts from the HTML templates?

File: giskard/scanner/issues.py [hunk 1]
```diff
@@ -25,6 +27,70 @@ class IssueGroup:
...
+            ExampleManager: current object with max number of elements set to n
+        """
+        self._max_num = n
+        return self
+
+    def __len__(self):
+        return len(self._examples)
+
+    def len(self):
+        return self.__len__()
```
Comment (rabah-khalek) [inline giskard/scanner/issues.py:69]: why do you need this?

File: giskard/scanner/issues.py [hunk 2]
```diff
@@ -25,6 +27,70 @@ class IssueGroup:
...
+        """
+        Add examples to the example manager
+
+        Args:
+            example (Any): new example to be added
+        """
+        if isinstance(example, list):
+            self._examples += example
+        else:
+            self._examples.append(example)
```
Comment (rabah-khalek) [inline giskard/scanner/issues.py:50]: I would make this abstract, since the passed `examples` can be `Any`

File: giskard/scanner/issues.py [hunk 3]
```diff
@@ -41,6 +107,8 @@ def __init__(
         features: Optional[List[str]] = None,
         tests=None,
         taxonomy: List[str] = None,
+        example_manager: Optional[ExampleManager] = ExampleManagerDataFrame,
+        display_warnings: Optional[bool] = True,
```
Comment (rabah-khalek) [inline giskard/scanner/issues.py:111]: let's call this `display_footer_info` instead

**Review by rabah-khalek — CHANGES_REQUESTED** [review]

File: giskard/scanner/issues.py [hunk 4]
```diff
@@ -25,6 +26,44 @@ class IssueGroup:
...
+
+class DataFrameScanExamples(ScanExamples):
+    """
+    ScanExamples class to manage examples from pandas dataframes
+    """
+
+    def __init__(self, examples: pd.DataFrame = None):
+        if examples is None:
+            examples = pd.DataFrame()
+        self.examples = examples
```
Comment (rabah-khalek) [inline giskard/scanner/issues.py:55]:
```suggestion
        self.examples = pd.DataFrame() if examples is None else examples
```

File: giskard/scanner/issues.py [hunk 5]
```diff
@@ -41,6 +80,8 @@ def __init__(
         features: Optional[List[str]] = None,
         tests=None,
         taxonomy: List[str] = None,
+        example_manager: Optional[ScanExamples] = None,
```
Comment (rabah-khalek) [inline giskard/scanner/issues.py:83]:
```suggestion
        scan_examples: Optional[ScanExamples] = None,
```

File: giskard/scanner/issues.py [hunk 6]
```diff
@@ -76,6 +117,10 @@ def __init__(
         taxonomy : Optional[str]
             List of taxonomy machine tags, in MISP format. A machine tag is composed of a namespace (MUST), a predicate
             (MUST) and an (OPTIONAL) value, like ``namespace:predicate:value``.
+        example_manager : Optional[ScanExamples]
```
Comment (rabah-khalek) [inline giskard/scanner/issues.py:120]:
```suggestion
        scan_examples : Optional[ScanExamples]
```

File: giskard/scanner/issues.py [hunk 7]
```diff
@@ -90,6 +135,12 @@ def __init__(
         self._features = features
         self._tests = tests
         self.taxonomy = taxonomy or []
+        self.display_footer_info = display_footer_info
+        if example_manager is None:
+            example_manager = DataFrameScanExamples()
+        self.example_manager = example_manager
+        if self._examples is not None:
+            self.example_manager.add_examples(self._examples)
```
Comment (rabah-khalek) [inline giskard/scanner/issues.py:143]:
```suggestion
        if scan_examples is None:
            scan_examples = DataFrameScanExamples()
        self.scan_examples = scan_examples
        if self._examples is not None:
            self.scan_examples.add_examples(self._examples)
```

File: giskard/scanner/issues.py [hunk 8]
```diff
@@ -129,16 +180,11 @@ def description(self):
...
-            return self._examples.head(n)
-        return pd.DataFrame()
-
-    def add_examples(self, examples: pd.DataFrame):
-        if self._examples is None:
-            self._examples = examples
-        else:
-            self._examples = pd.concat([self._examples, examples])
+    def examples(self, n=3) -> Any:
+        return self.example_manager.head(n)
```
Comment (rabah-khalek) [inline giskard/scanner/issues.py:184]:
```suggestion
        return self.scan_examples.head(n)
```

File: giskard/scanner/issues.py [hunk 9]
```diff
@@ -25,6 +26,42 @@ class IssueGroup:
...
+    ScanExamples class to manage examples from pandas dataframes
+    """
+
+    def __init__(self, examples: pd.DataFrame = None):
+        self.examples = pd.DataFrame() if examples is None else examples
+
+    def add_examples(self, example):
+        self.examples = pd.concat([self.examples, example])
+
+    def head(self, n):
```
Comment (rabah-khalek) [inline giskard/scanner/issues.py:58]:
```suggestion
    def head(self, n: int):
```

File: giskard/scanner/issues.py [hunk 10]
```diff
@@ -25,6 +26,42 @@ class IssueGroup:
     description: str
 
 
+class ScanExamples(ABC):
```
Comment (Hartorn) [inline giskard/scanner/issues.py:29]: Are we using naming with s for plural elsewhere ?

File: giskard/scanner/issues.py [hunk 11]
```diff
@@ -25,6 +26,42 @@ class IssueGroup:
...
+    """
+    Abstract class to manage examples from different data types.
+    """
+
+    @abstractmethod
+    def add_examples(self, example: Any):
```
Comment (Hartorn) [inline giskard/scanner/issues.py:35]: naming must be consistent between function & argument (is it one or several arguments we are adding ?)

if it's several, i would go with `extend`to copy existing naming in python

Comment (Hartorn) [reply giskard/scanner/issues.py:35]: Also, missing doc string

File: giskard/scanner/issues.py [hunk 12]
```diff
@@ -25,6 +26,42 @@ class IssueGroup:
     description: str
 
 
+class ScanExamples(ABC):
+    """
+    Abstract class to manage examples from different data types.
```
Comment (Hartorn) [inline giskard/scanner/issues.py:31]: Could define what's the user of the examples also

File: giskard/scanner/issues.py [hunk 13]
```diff
@@ -25,6 +26,42 @@ class IssueGroup:
...
+    def to_html(self):
+        ...
+
+
+class DataFrameScanExamples(ScanExamples):
+    """
+    ScanExamples class to manage examples from pandas dataframes
+    """
+
+    def __init__(self, examples: pd.DataFrame = None):
```
Comment (Hartorn) [inline giskard/scanner/issues.py:52]: Type should be Optional[pd.DataFrame] then

File: giskard/scanner/issues.py [hunk 14]
```diff
@@ -25,6 +26,42 @@ class IssueGroup:
...
+        self.examples = pd.DataFrame() if examples is None else examples
```
Comment (Hartorn) [inline giskard/scanner/issues.py:53]: Should we check if examples is a dataframe ?

File: giskard/scanner/issues.py [hunk 15]
```diff
@@ -25,6 +26,42 @@ class IssueGroup:
...
+class ScanExamples(ABC):
+    """
+    Abstract class to manage examples from different data types.
+    """
+
+    @abstractmethod
+    def add_examples(self, example: Any):
+        ...
+
+    @abstractmethod
```
Comment (Hartorn) [inline giskard/scanner/issues.py:38]: Same

File: giskard/scanner/issues.py [hunk 16]
```diff
@@ -25,6 +26,42 @@ class IssueGroup:
...
+    def head(self, n: int):
+        ...
+
+    @abstractmethod
```
Comment (Hartorn) [inline giskard/scanner/issues.py:42]: Same

File: giskard/scanner/issues.py [hunk 17]
```diff
@@ -90,6 +133,10 @@ def __init__(
         self._features = features
         self._tests = tests
         self.taxonomy = taxonomy or []
+        self.display_footer_info = display_footer_info
+        self.scan_examples = DataFrameScanExamples() if scan_examples is None else scan_examples
```
Comment (Hartorn) [inline giskard/scanner/issues.py:137]: I don't like we are defaulting to dataframe, but that's ok I guess

File: giskard/scanner/issues.py [hunk 18]
```diff
@@ -90,6 +133,10 @@ def __init__(
...
+        if self._examples is not None:
```
Comment (Hartorn) [inline giskard/scanner/issues.py:138]: Do we still need to assign something to self._examples ?
We could just use the arguments here

File: giskard/scanner/issues.py [hunk 19]
```diff
@@ -41,6 +78,8 @@ def __init__(
         features: Optional[List[str]] = None,
         tests=None,
         taxonomy: List[str] = None,
+        scan_examples: Optional[ScanExamples] = None,
+        display_footer_info: Optional[bool] = True,
```
Comment (Hartorn) [inline giskard/scanner/issues.py:82]: It's not optional bool, just bool with default value

**Review by Hartorn — CHANGES_REQUESTED** [review]
Miss click

File: giskard/scanner/issues.py (hunk 14 above)
Comment (rabah-khalek) [reply giskard/scanner/issues.py:53]: sure, we can add the `@configured_validate_arguments` on the `__init__`

File: giskard/scanner/issues.py (hunk 10 above)
Comment (bmalezieux) [reply giskard/scanner/issues.py:30]: This is used to manage several examples at once, so I guess the plural is better

**Pull request merged**